analyzers/rolaguard_base_analyzer/DeviceIdentifier.py

```diff
@@ -0,0 +1,202 @@
+
+
+    def get_deveui_from_session(self, packet):
+        """
+        Looks for the device_session of the packet, from this object gets the
+        device_id and from this finally gets the dev_eui. If one step fails raises an exception.
+        """
+        device_session = DeviceSession.find_with(
+            dev_addr=packet.dev_addr,
+            data_collector_id=packet.data_collector_id
+            )
+        device = Device.get(device_session.device_id)
+        packet.dev_eui = device.dev_eui
+        return packet
+
+
+    def guess_deveui_from_jrs(self, packet):
+        """
+        Guess the dev_eui using the rssi, lsnr, gateway and machine learning.
+        """
+        if packet.f_count > 4:
+            # The data packet has a high counter, the JR was missed.
+            return packet
+        if packet.rssi is None or packet.lsnr is None:
+            # Data needed to guess the device is missing.
+            return packet
+        join_request = [jr for jr in self.jrs[packet.data_collector_id] if jr["ja_counter"]>0]
+        if len(join_request) < 1:
+            # No JR registered in the last jr_expiration_time seconds.
+            return packet
+
+        ds = pd.DataFrame(self.jrs[packet.data_collector_id])
+
+        deveui_encoder = LabelEncoder().fit(ds["dev_eui"])
+        if len(deveui_encoder.classes_) < 2:
+            # All JRs registered are from the same device, the packet must be from this device
+            packet.dev_eui = deveui_encoder.classes_[0]
+            return packet
+
+        # Some data preprocessing
+        gateway_encoder = OneHotEncoder().fit(np.array(ds["gateway"]).reshape(-1,1))
+        numeric_scaler = RobustScaler().fit(ds[["rssi", "lsnr"]])
+        label_encoder = LabelEncoder().fit(ds["dev_eui"])
+        X = np.concatenate([
+                gateway_encoder.transform(np.array(ds["gateway"]).reshape(-1,1)).toarray(),
+                numeric_scaler.transform(ds[["rssi", "lsnr"]]),
+            ], axis=1
+        )
+        Y = label_encoder.transform(ds["dev_eui"])
+        
+        # Trains a Support Vector Classifier with saved JRs.
+        # The hiper-parameters were optimized with grid search.
+        model = SVC(C=0.5, gamma=1.0, class_weight='balanced').fit(X, Y)
+        
+        # Makes a prediction over the actual packet.
+        X = np.concatenate([
+                gateway_encoder.transform(np.array(packet.gateway).reshape(1,1)).toarray(),
+                numeric_scaler.transform(np.array([[packet.rssi, packet.lsnr]]))
+            ], axis=1
+        )
+        predicted_dev_eui = label_encoder.inverse_transform(model.predict(X))[0]
+        packet.dev_eui = predicted_dev_eui
+
+        # All registered JRs from the predicted device are deleted, since it is probabily linked to a session.
+        self.del_jrs_with(data_collector_id=packet.data_collector_id, dev_eui=predicted_dev_eui)
+        try: # Tries to unlink the device and the session (if any)
```

Comment:
"Tries to unlink" should be "Tries to link"

Reply:
Fixed!